apps/prairielearn/src/lib/groups.ts

```diff
@@ -200,79 +265,96 @@ export async function joinGroup(
   const splitJoinCode = fullJoinCode.split('-');
   if (splitJoinCode.length !== 2 || splitJoinCode[1].length !== 4) {
     // the join code input by user is not valid (not in format of groupname+4-character)
-    flash('error', 'The join code has an incorrect format');
+    throw new Error('The join code has an incorrect format');
     return;
   }
 
   const groupName = splitJoinCode[0];
   const joinCode = splitJoinCode[1].toUpperCase();
 
-  // This is a best-effort check to produce a nice error message. Even if this
-  // fails due to a race condition, the `group_users_insert` sproc below will
-  // validate that the user isn't already in a group.
-  const existingGroupId = await getGroupId(assessmentId, userId);
-  if (existingGroupId != null) {
-    flash('error', 'You are already in another group.');
-    return;
-  }
-
   try {
-    await sqldb.callAsync('group_users_insert', [
-      assessmentId,
-      userId,
-      authnUserId,
-      groupName,
-      joinCode,
-    ]);
+    await sqldb.runInTransactionAsync(async () => {
+      const group = await sqldb.queryOptionalRow(
+        sql.select_group_by_name,
+        { group_name: groupName, assessment_id: assessmentId },
+        GroupSchema,
+      );
+      if (group == null || group.join_code !== joinCode) {
+        throw new Error('Group does not exist.');
+      }
+      await addUserToGroup(assessmentId, group.id, userId, authnUserId, true);
+    });
   } catch (err) {
-    flash(
-      'error',
-      `Failed to join the group with join code ${fullJoinCode}. It is already full or does not exist. Please try to join another one.`,
-    );
+    throw new Error(`Cannot join group "${fullJoinCode}": ${err.message}`);
   }
 }
 
 export async function createGroup(
   groupName: string,
   assessmentId: string,
-  userId: string,
+  userIds: string[],
   authnUserId: string,
 ): Promise<void> {
   if (groupName.length > 30) {
-    flash('error', 'The group name is too long. Use at most 30 alphanumerical characters.');
-    return;
+    throw new Error('The group name is too long. Use at most 30 alphanumerical characters.');
```

Comment:
I won't block on this, but: the shift from using flash messages to `throw new Error(...)` means that any failure to create groups will show up as 500 errors in Sentry and will register in our error metrics in CloudWatch. I'd much prefer to attach a suitable 4XX error code to these sorts of "expected" errors and reserve 500s for when unexpected things go wrong.

Reply:
Note that the exceptions are caught in all calls to these functions, so they don't reach the 500 error (as far as I can tell). I shifted to using exceptions so that the error can be sent to either flash or the job logger depending on context. An alternative is to keep using `flash` and have the job cases (auto-groups and upload) read from flash, but I find that a bit weird. Alternatively I can also have it return the error cases, similar to issues returned by freeform.

Reply:
I very much agree that the jobs shouldn't be reading from `req.flash`. My concern is not with the move from `flash()` to thrown errors, it's the fact that at multiple levels here we fail to differentiate between expected and unexpected errors. I think returning something like a status enum would be an OK approach, as would throwing a subclass of `Error` that we could specifically handle at callsites where we want to differentiate between expected/unexpected errors.

There's a separate comment about code that looks like `.catch((err) => flash('error', err.message));`, which should be considered as you decide what to do here.

Reply:
I can think of two alternatives to throwing errors: returning a list of errors (a la freeform), or adding a callback argument for error reporting. Any preference?

Reply:
I don't very much like either of those. For the former: when would we ever expect to produce multiple errors here? For the latter: that makes control flow difficult to reason about and implement correctly.

Reply:
Maybe return `{ success: true } | { success: false; error: Error }`?

Reply:
Replaced with a custom error class in 65872d4f3. I used an error since the transaction needs to be rolled back anyway in some cases.